sda/router.go

```diff
@@ -0,0 +1,543 @@
+func (t *TCPRouter) handleConn(c network.SecureConn) {
+	address := c.Remote()
+	for {
+		ctx := context.TODO()
+		packet, err := c.Receive(ctx)
+		// So the receiver can know about the error
+		packet.SetError(err)
+		packet.From = address
+		log.Lvl5(t.workingAddress, "Got message", packet)
+		if err != nil {
+			t.closingMut.Lock()
+			log.Lvlf4("%+v got error (%+s) while receiving message (isClosing=%+v)",
+				t.serverIdentity.First(), err, t.isClosing)
+			t.closingMut.Unlock()
+			if err == network.ErrClosed || err == network.ErrEOF || err == network.ErrTemp {
+				log.Lvl4(t.serverIdentity.First(), c.Remote(), "quitting handleConn for-loop", err)
+				t.closeConnection(c)
+				return
+			}
+			log.Error(t.serverIdentity.Addresses, "Error with connection", address, "=>", err)
+		} else {
+			t.closingMut.Lock()
+			if !t.isClosing {
+				log.Lvl5(t.workingAddress, "Send message to networkChan", len(t.networkChan))
+				t.networkChan <- packet
+			}
+			t.closingMut.Unlock()
+		}
+	}
+}
+
+// registerConnection registers an ServerIdentity for a new connection, mapped with the
+// real physical address of the connection and the connection itself
+// it locks (and unlocks when done):  networkLock
+func (t *TCPRouter) registerConnection(c network.SecureConn) {
+	log.Lvl4(t.serverIdentity.First(), "registers", c.ServerIdentity().First())
+	t.networkLock.Lock()
+	defer t.networkLock.Unlock()
+	id := c.ServerIdentity()
+	_, okc := t.connections[id.ID]
+	if okc {
+		log.Lvl3("Connection already registered", okc)
+	}
+	t.connections[id.ID] = c
+}
+
+// Tx to implement monitor/CounterIO
+func (t *TCPRouter) Tx() uint64 {
+	return t.host.Tx()
+}
+
+// Rx to implement monitor/CounterIO
+func (t *TCPRouter) Rx() uint64 {
+	return t.host.Rx()
+}
+
+// Address is the address where this host is listening
+func (t *TCPRouter) Address() string {
+	return t.workingAddress
+}
+
+// GetStatus is a function that returns the status report of the server.
+func (t *TCPRouter) GetStatus() Status {
+	m := make(map[string]string)
+	nbr := len(t.connections)
+	remote := make([]string, nbr)
+	iter := 0
+	var rx uint64
+	var tx uint64
+	for _, c := range t.connections {
+		remote[iter] = c.Remote()
+		rx += c.Rx()
+		tx += c.Tx()
+		iter = iter + 1
+	}
+	m["Connections"] = strings.Join(remote, "\n")
+	m["Host"] = t.Address()
+	m["Total"] = strconv.Itoa(nbr)
+	m["Packets_Received"] = strconv.FormatUint(rx, 10)
+	m["Packets_Sent"] = strconv.FormatUint(tx, 10)
+	return m
+}
+
+func (t *TCPRouter) connection(e *network.ServerIdentity) network.SecureConn {
+	t.networkLock.RLock()
+	defer t.networkLock.RUnlock()
+	c, _ := t.connections[e.ID]
+	return c
+}
+
+// MOCKING NETWORK ROUTER
+// localRelay defines the basic functionalities such as sending and
+// receiving a message, locally. It is implemented by LocalRouter and
+// localClient so both a Router and a Client can be emulated locally without
+// opening any real connections.
+type localRelay interface {
```

Comment:
I wanted so at the beginning but then we can't use mocking network outside of `sda/` anymore... which is also the whole point of this PR. Is there a way for test packages to import other test packages and so on ? If possible then yes. I haven't looked it through. The whole `local.go` file could also be put into a test file then.